blocklib/limesdr/limesdr/limesdr_cpu.h

```diff
@@ -0,0 +1,588 @@
+            LMS_GetSampleRateRange(d_device->handle, LMS_CH_RX, &sr_range);
+        if (sample_range_rc == LMS_SUCCESS) {
+            d_logger->info("Sample rate range: Min: {} Max: {}, Step: {}",
+                           sr_range.min,
+                           sr_range.max,
+                           sr_range.step);
+        }
+
+        const auto sample_rate = std::get<double>(*this->param_sample_rate);
+        const auto set_samp_rc = LMS_SetSampleRate(d_device->handle, sample_rate, 0);
+        if (set_samp_rc != LMS_SUCCESS) {
+            throw std::runtime_error(
+                fmt::format("Could not set sample rate ({})", sample_rate));
+        }
+
+        d_logger->info("Set sample rate {}", sample_rate);
+
+        static constexpr int MAX_CHANNELS = 2;
+        if (num_device_channels > MAX_CHANNELS) {
+            d_logger->warn("Device has more channels than block allows! ({} vs {})",
+                           num_device_channels,
+                           MAX_CHANNELS);
+        }
+
+        const auto n_channels =
+            static_cast<std::size_t>(std::min(num_device_channels, MAX_CHANNELS));
+        d_device->streams.resize(n_channels);
+
+        const auto enabled_channels =
+            pmtv::get_vector<std::size_t>(*this->param_enabled_channels);
+        const auto trigger_mask = std::get<int>(*this->param_timing_trigger_mask);
+
+        for (std::size_t channel_idx = 0; channel_idx < n_channels; channel_idx++) {
+            const auto enable = std::find(enabled_channels.begin(),
+                                          enabled_channels.end(),
+                                          channel_idx) != enabled_channels.end();
+
+            const auto enable_channel_rc =
+                LMS_EnableChannel(d_device->handle, LMS_CH_RX, channel_idx, enable);
+            if (enable_channel_rc != LMS_SUCCESS) {
+                throw std::runtime_error(
+                    fmt::format("Could not enable/disable channel {} (enable: {})",
+                                channel_idx,
+                                enable));
+            }
+
+            d_logger->info("Set channel {} (enabled={})", channel_idx, enable);
+
+            if (!enable) {
+                continue;
+            }
+
+            const auto driver_buffer_size =
+                std::get<std::size_t>(*this->param_driver_buffer_size);
+
+            lms_stream_t stream_cfg{ .isTx = LMS_CH_RX,
+                                     .channel = static_cast<uint32_t>(channel_idx),
+                                     .fifoSize =
+                                         static_cast<uint32_t>(driver_buffer_size),
+                                     .throughputVsLatency =
+                                         0.5, // TODO what is wanted here? 0 is lowest
+                                              // latency, 1 highest throughput
+                                     .dataFmt = lms_stream_t::LMS_FMT_I12 };
```

Comment:
Is this the right format? (I16 doesn't help with the trigger bits though)

Reply:
The `.dataFmt = lms_stream_t::LMS_FMT_I12` looks a bit strange. The FPGA returns 16 values where the 4 LSBs contain the GPIO states... i.e. w/o trigger all samples should be larger than 15 and/or 'sample & 0xF == 0'. :thinking: 

Could you check for that? Thanks in advance.

Reply:
The options are

`
LMS_FMT_F32=0,    ///<32-bit floating point
LMS_FMT_I16,      ///<16-bit integers
LMS_FMT_I12       ///<12-bit integers stored in 16-bit variables
`

(always two values to for a sample, for I and Q). My thinking, after starting with LMS_FMT_I16, was: Maybe they use the 12bit in 16bit option, and use the remaining 4 bits for the digital ports. However, with LMS_FMT_I16 (and LMS_FMT_I12 isn't much different), I see values like:

`
          111111 1111111011110000
1111111011001111      10011111111
1111110111111111 1111110011111111
1111111011101111 1111111011111111
`

So I don't see any bits that are consistently 0 or consistently 1. I wonder if the splitting into I and Q components hide the original digital bit information?

Reply:
Note to self: LMS_FMT_I16 with two enabled channels seems the least complex case, the data read from the FPGA is passed through unmodified (?). I can't identify the digital bits (which I expect to be always 0) in the data from the FPGA either though, i.e. the input data passed in here: https://github.com/myriadrf/LimeSuite/blob/64c6155e62ffa7f44b812efcd865405977b818b2/src/FPGA_common/FPGA_common.cpp#L427

Reply:
LMS_FMT_I16 , digital bits seem to work fine now in Alex' setup, and we get consistent albeit weird values (like, 0b0001) at least.